parlai/nn/lr_scheduler.py

```diff
@@ -0,0 +1,357 @@
+        self.warmup_scheduler, if appropriate.
+
+        :param opt opt: Arguments received by torch_agent
+        :param optimizer optimizer: Optimizer being used for training. May be
+            wrapped in fp16_optimizer_wrapper depending on whether fp16 is used
+        :param state_dict states: Possible state_dict provided by model
+            checkpoint, for restoring LR state
+        :param bool hard_reset: If true, the LR scheduler should ignore the
+            state dictionary.
+        """
+
+        patience = opt.get('lr_scheduler_patience', 3)
+        decay = opt.get('lr_scheduler_decay', 0.5)
+        warmup_updates = opt.get('warmup_updates', -1)
+        max_lr_steps = opt.get('max_lr_steps', -1)
+
+        if opt.get('lr_scheduler') == 'none':
+            scheduler = None
+        elif decay == 1.0:
+            warn_once(
+                "Your LR decay is set to 1.0. Assuming you meant you wanted "
+                "to disable learning rate scheduling. Adjust --lr-scheduler-decay "
+                "if this is not correct."
+            )
+            scheduler = None
+        elif opt.get('lr_scheduler') == 'reduceonplateau':
+            scheduler = ReduceOnPlateauLRScheduler(
+                optimizer, states, hard_reset, patience, decay, warmup_updates
+            )
+        elif opt.get('lr_scheduler') == 'fixed':
+            scheduler = FixedLRScheduler(
+                optimizer, states, hard_reset, patience, decay, warmup_updates
+            )
+        elif opt.get('lr_scheduler') == 'invsqrt':
+            scheduler = InvSqrtLRScheduler(
+                optimizer,
+                states,
+                hard_reset,
+                patience,
+                decay,
+                warmup_updates,
+                max_lr_steps,
+            )
+        elif opt.get('lr_scheduler') == 'cosine':
+            scheduler = CosineLRScheduler(
+                optimizer,
+                states,
+                hard_reset,
+                patience,
+                decay,
+                warmup_updates,
+                max_lr_steps,
+            )
+        elif opt.get('lr_scheduler') == 'linear':
+            scheduler = LinearLRScheduler(
+                optimizer,
+                states,
+                hard_reset,
+                patience,
+                decay,
+                warmup_updates,
+                max_lr_steps,
+            )
+        else:
+            raise ValueError(
+                "Don't know what to do with lr_scheduler '{}'".format(
+                    opt.get('lr_scheduler')
+                )
+            )
+
+        # time to load LR state from the checkpoint, if possible.
+        if (
+            # there is already an old LR scheduler saved on disk
+            states
+            and
+            # and the old LR scheduler is different
+            states.get('lr_scheduler_type') != opt['lr_scheduler']
```

Comment:
Nit: move the `and` to be on the next line so black isn't upset.

Reply:
? confused, isn't it already?

Reply:
`and states.get(‘lr_...`